File: tests/python/Test_Binding_Equivalence.py
# pytamm (Python) half of the binding-equivalence check.
#
# Runs a fixed set of tensor operations through the pytamm bindings and writes
# the results to disk under the `py_tamm_equiv` prefix.
#
# Part of a 3-file check proving the Python bindings produce numerically
# identical results to core C++ TAMM:
#   1. Test_Binding_Equivalence.cpp    -> writes `cpp_tamm_equiv` (same ops, C++)
#   2. Test_Binding_Equivalence.py     -> writes `py_tamm_equiv`  (this file)
#   3. Compare_Binding_Equivalence.py  -> reads both, asserts they match
# Run 1 and 2 (order-independent), then 3.

import argparse
import os
import sys

import pytamm as pt


def parse_args():
    p = argparse.ArgumentParser()
    p.add_argument("--prefix", default="py_tamm_equiv")
    args, _ = p.parse_known_args()
    return args


def nz_a(b):
    x, y = int(b[0]), int(b[1])
    return ((x + 2 * y) % 3) != 1


def nz_b(b):
    x, y = int(b[0]), int(b[1])
    return ((2 * x + y) % 3) != 2


def nz_d(b):
    x, y = int(b[0]), int(b[1])
    return x == y or ((x + y) % 2 == 0)


def nz_c(b):
    x, y = int(b[0]), int(b[1])
    return x <= y or ((x + y) % 2 == 0)


def dval(which, bid, p):
    x, y, q = int(bid[0]), int(bid[1]), int(p)

    if which == 0:
        n = 16 * x + 4 * y + q + 1
    elif which == 1:
        n = 8 * x - 6 * y + q + 3
    else:
        n = 5 * x + 7 * y + 2 * q + 1

    return n / 8.0


def zval(which, bid, p):
    x, y, q = int(bid[0]), int(bid[1]), int(p)

    if which == 0:
        re = 6 * x + 2 * y + q + 1
        im = 3 * x - 5 * y + q + 2
    else:
        re = -4 * x + 7 * y + q + 1
        im = 5 * x + y - q - 1

    return complex(re / 8.0, im / 8.0)


def fill_double_tensor(tensor, which):
    def fill(blockid, buf):
        for p in range(buf.size):
            buf[p] = dval(which, blockid, p)

    pt.fill_sparse_tensor_double(tensor, fill)


def fill_complex_tensor(tensor, which):
    def fill(blockid, buf):
        for p in range(buf.size):
            buf[p] = zval(which, blockid, p)

    pt.fill_sparse_tensor_complex_double(tensor, fill)


def make_spaces():
    ispace = pt.IndexSpace(pt.range(6))
    return pt.TiledIndexSpace(ispace, 2)


def make_tensors(x):
    info_a = pt.TensorInfo([x, x], nz_a)
    info_b = pt.TensorInfo([x, x], nz_b)
    info_d = pt.TensorInfo([x, x], nz_d)
    info_c = pt.TensorInfo([x, x], nz_c)

    A = pt.TensorDouble([x, x], info_a)
    B = pt.TensorDouble([x, x], info_b)
    D = pt.TensorDouble([x, x], info_d)
    C = pt.TensorDouble([x, x], info_c)
    R = pt.TensorDouble([x, x], info_c)

    ZA = pt.TensorComplexDouble([x, x], info_a)
    ZB = pt.TensorComplexDouble([x, x], info_b)
    ZC = pt.TensorComplexDouble([x, x], info_c)

    return A, B, D, C, R, ZA, ZB, ZC


def run_scheduler_double(ec, i, j, k, A, B, D, C):
    sch = pt.Scheduler(ec)
    sch(C(i, j), "=", 0.0)
    sch(C(i, j), "+=", 1.5, A(i, k), B(k, j))
    sch(C(i, j), "-=", 0.25, D(i, j))
    sch.execute(pt.ExecutionHW.CPU, False)


def run_scheduler_complex(ec, i, j, k, ZA, ZB, ZC):
    alpha = complex(0.5, -0.25)
    beta = complex(-0.125, 0.25)

    sch = pt.Scheduler(ec)
    sch(ZC(i, j), "=", 0.0 + 0.0j)
    sch(ZC(i, j), "+=", alpha, ZA(i, k), ZB(k, j))
    sch(ZC(i, j), "+=", beta, ZA(i, j))
    sch.execute(pt.ExecutionHW.CPU, False)


def run_opmin_double(ec, i, j, k, A, B, D, R):
    pt.Scheduler(ec)(R(i, j), "=", 0.0).execute(pt.ExecutionHW.CPU, False)

    rhs = (
        1.25 * pt.as_op(A(i, k)) * pt.as_op(B(k, j))
        + (-0.5) * pt.as_op(D(i, j))
    )

    R(i, j).update(rhs)

    st = pt.SymbolTable()
    pt.register_symbols(st, A=A, B=B, D=D, R=R, i=i, j=j, k=k)

    sch = pt.Scheduler(ec)
    ex = pt.OpExecutor(sch, st)
    ex.opmin_execute(R)


def run(prefix):
    pg = pt.ProcGroup.create_world_coll()
    ec = pt.ExecutionContext(pg, pt.DistributionKind.nw, pt.MemoryManagerKind.ga)

    if ec.pg_rank() == 0:
        dirname = os.path.dirname(os.path.abspath(prefix))
        if dirname:
            os.makedirs(dirname, exist_ok=True)

    pg.barrier()

    x = make_spaces()
    i, j, k = x.labels("all", None, 3)

    A, B, D, C, R, ZA, ZB, ZC = make_tensors(x)

    pt.TensorDouble.allocate(ec, A, B, D, C, R)
    pt.TensorComplexDouble.allocate(ec, ZA, ZB, ZC)

    fill_double_tensor(A, 0)
    fill_double_tensor(B, 1)
    fill_double_tensor(D, 2)
    fill_complex_tensor(ZA, 0)
    fill_complex_tensor(ZB, 1)

    run_scheduler_double(ec, i, j, k, A, B, D, C)
    run_opmin_double(ec, i, j, k, A, B, D, R)
    run_scheduler_complex(ec, i, j, k, ZA, ZB, ZC)

    ec.flush_and_sync()

    pt.write_to_disk(C, prefix + "_sched_d", True, False, 0)
    pt.write_to_disk(R, prefix + "_opmin_d", True, False, 0)
    pt.write_to_disk(ZC, prefix + "_sched_z", True, False, 0)

    ec.flush_and_sync()
    pg.barrier()

    if ec.pg_rank() == 0:
        print(f"Wrote {prefix}_sched_d, {prefix}_opmin_d, {prefix}_sched_z")

    for t in (ZC, ZB, ZA, R, C, D, B, A):
        t.deallocate()


def main():
    args = parse_args()
    pt.initialize(sys.argv, False)

    try:
        run(args.prefix)
    finally:
        pt.finalize()


if __name__ == "__main__":
    main()

File: tests/tamm/Test_Binding_Equivalence.cpp
// Core C++ TAMM half of the binding-equivalence check.
//
// Runs a fixed set of tensor operations through core C++ TAMM and writes the
// results to disk under the `cpp_tamm_equiv` prefix.
//
// Part of a 3-file check proving the pytamm (Python) bindings produce
// numerically identical results to core C++ TAMM:
//   1. Test_Binding_Equivalence.cpp    -> writes `cpp_tamm_equiv` (this file)
//   2. Test_Binding_Equivalence.py     -> writes `py_tamm_equiv`  (same ops, Python)
//   3. Compare_Binding_Equivalence.py  -> reads both, asserts they match
// Run 1 and 2 (order-independent), then 3.

#include "tamm/op_executor.hpp"
#include "tamm/opmin.hpp"
#include "tamm/tamm.hpp"

#include <complex>
#include <exception>
#include <iostream>
#include <string>

using namespace tamm;

namespace {

std::string arg_value(int argc, char** argv, const std::string& key, std::string def) {
  for(int i = 1; i < argc; ++i) {
    const std::string a = argv[i];
    if(a == key && i + 1 < argc) return argv[i + 1];
    const std::string p = key + "=";
    if(a.rfind(p, 0) == 0) return a.substr(p.size());
  }
  return def;
}

int bi(const IndexVector& b, size_t i) { return static_cast<int>(b[i]); }

bool nz_a(const IndexVector& b) { return ((bi(b, 0) + 2 * bi(b, 1)) % 3) != 1; }

bool nz_b(const IndexVector& b) { return ((2 * bi(b, 0) + bi(b, 1)) % 3) != 2; }

bool nz_d(const IndexVector& b) {
  const int x = bi(b, 0);
  const int y = bi(b, 1);
  return x == y || ((x + y) % 2 == 0);
}

bool nz_c(const IndexVector& b) {
  const int x = bi(b, 0);
  const int y = bi(b, 1);
  return x <= y || ((x + y) % 2 == 0);
}

TensorInfo make_info2(const TiledIndexSpace& x, NonZeroCheck check) {
  return TensorInfo{TiledIndexSpaceVec{x, x}, check};
}

double dval(int which, const IndexVector& b, size_t p) {
  const long long x = bi(b, 0);
  const long long y = bi(b, 1);
  const long long q = static_cast<long long>(p);

  long long n = 0;
  if(which == 0) n = 16 * x + 4 * y + q + 1;
  if(which == 1) n = 8 * x - 6 * y + q + 3;
  if(which == 2) n = 5 * x + 7 * y + 2 * q + 1;

  return static_cast<double>(n) / 8.0;
}

std::complex<double> zval(int which, const IndexVector& b, size_t p) {
  const long long x = bi(b, 0);
  const long long y = bi(b, 1);
  const long long q = static_cast<long long>(p);

  long long re = 0;
  long long im = 0;

  if(which == 0) {
    re = 6 * x + 2 * y + q + 1;
    im = 3 * x - 5 * y + q + 2;
  }
  else {
    re = -4 * x + 7 * y + q + 1;
    im = 5 * x + y - q - 1;
  }

  return {static_cast<double>(re) / 8.0, static_cast<double>(im) / 8.0};
}

void fill_d(Tensor<double>& t, int which) {
  fill_sparse_tensor<double>(t, [which](const IndexVector& bid, span<double> buf) {
    for(size_t p = 0; p < buf.size(); ++p) buf[p] = dval(which, bid, p);
  });
}

void fill_z(Tensor<std::complex<double>>& t, int which) {
  fill_sparse_tensor<std::complex<double>>(
    t, [which](const IndexVector& bid, span<std::complex<double>> buf) {
      for(size_t p = 0; p < buf.size(); ++p) buf[p] = zval(which, bid, p);
    });
}

void run_scheduler_double(ExecutionContext& ec, const TiledIndexLabel& i, const TiledIndexLabel& j,
                          const TiledIndexLabel& k, Tensor<double>& A, Tensor<double>& B,
                          Tensor<double>& D, Tensor<double>& C) {
  Scheduler{ec}(C(i, j) = 0.0)(C(i, j) += 1.5 * A(i, k) * B(k, j))(C(i, j) -= 0.25 * D(i, j))
    .execute(ExecutionHW::CPU, false);
}

void run_scheduler_complex(ExecutionContext& ec, const TiledIndexLabel& i, const TiledIndexLabel& j,
                           const TiledIndexLabel& k, Tensor<std::complex<double>>& ZA,
                           Tensor<std::complex<double>>& ZB, Tensor<std::complex<double>>& ZC) {
  const std::complex<double> alpha{0.5, -0.25};
  const std::complex<double> beta{-0.125, 0.25};

  Scheduler{ec}(ZC(i, j) = std::complex<double>{0.0, 0.0})(ZC(i, j) += alpha * ZA(i, k) * ZB(k, j))(
    ZC(i, j) += beta * ZA(i, j))
    .execute(ExecutionHW::CPU, false);
}

void run_opmin_double(ExecutionContext& ec, const TiledIndexLabel& i, const TiledIndexLabel& j,
                      const TiledIndexLabel& k, Tensor<double>& A, Tensor<double>& B,
                      Tensor<double>& D, Tensor<double>& R) {
  Scheduler{ec}(R(i, j) = 0.0).execute(ExecutionHW::CPU, false);

  auto rhs = (1.25 * tamm::new_ops::LTOp{A(i, k)} * tamm::new_ops::LTOp{B(k, j)}) +
             (-0.5 * tamm::new_ops::LTOp{D(i, j)});

  R(i, j).update(rhs);

  SymbolTable symbol_table;
  TAMM_REGISTER_SYMBOLS(symbol_table, A, B, D, R);
  TAMM_REGISTER_SYMBOLS(symbol_table, i, j, k);

  Scheduler  sch{ec};
  OpExecutor op_exec{sch, symbol_table};

  op_exec.opmin_execute(R);
}

template<typename T>
void safe_deallocate(Tensor<T>& t) {
  if(t.is_allocated()) t.deallocate();
}

} // namespace

int main(int argc, char** argv) {
  tamm::initialize(argc, argv, false);

  int rc = 0;

  try {
    const std::string prefix = arg_value(argc, argv, "--prefix", "cpp_tamm_equiv");

    ProcGroup        pg = ProcGroup::create_world_coll();
    ExecutionContext ec{pg, DistributionKind::nw, MemoryManagerKind::ga};

    IndexSpace      is{range(6)};
    TiledIndexSpace x{is, 2};

    auto [i, j, k] = x.labels<3>("all");

    const auto info_a = make_info2(x, NonZeroCheck{nz_a});
    const auto info_b = make_info2(x, NonZeroCheck{nz_b});
    const auto info_d = make_info2(x, NonZeroCheck{nz_d});
    const auto info_c = make_info2(x, NonZeroCheck{nz_c});

    Tensor<double> A{TiledIndexSpaceVec{x, x}, info_a};
    Tensor<double> B{TiledIndexSpaceVec{x, x}, info_b};
    Tensor<double> D{TiledIndexSpaceVec{x, x}, info_d};
    Tensor<double> C{TiledIndexSpaceVec{x, x}, info_c};
    Tensor<double> R{TiledIndexSpaceVec{x, x}, info_c};

    Tensor<std::complex<double>> ZA{TiledIndexSpaceVec{x, x}, info_a};
    Tensor<std::complex<double>> ZB{TiledIndexSpaceVec{x, x}, info_b};
    Tensor<std::complex<double>> ZC{TiledIndexSpaceVec{x, x}, info_c};

    A.allocate(&ec);
    B.allocate(&ec);
    D.allocate(&ec);
    C.allocate(&ec);
    R.allocate(&ec);
    ZA.allocate(&ec);
    ZB.allocate(&ec);
    ZC.allocate(&ec);

    fill_d(A, 0);
    fill_d(B, 1);
    fill_d(D, 2);
    fill_z(ZA, 0);
    fill_z(ZB, 1);

    run_scheduler_double(ec, i, j, k, A, B, D, C);
    run_opmin_double(ec, i, j, k, A, B, D, R);
    run_scheduler_complex(ec, i, j, k, ZA, ZB, ZC);

    ec.flush_and_sync();

    write_to_disk<double>(C, prefix + "_sched_d", true, false, 0);
    write_to_disk<double>(R, prefix + "_opmin_d", true, false, 0);
    write_to_disk<std::complex<double>>(ZC, prefix + "_sched_z", true, false, 0);

    ec.flush_and_sync();
    pg.barrier();

    if(pg.rank() == 0) {
      std::cout << "Wrote " << prefix << "_sched_d, " << prefix << "_opmin_d, " << prefix
                << "_sched_z\n";
    }

    safe_deallocate(ZC);
    safe_deallocate(ZB);
    safe_deallocate(ZA);
    safe_deallocate(R);
    safe_deallocate(C);
    safe_deallocate(D);
    safe_deallocate(B);
    safe_deallocate(A);
  } catch(const std::exception& ex) {
    std::cerr << "C++ TAMM equivalence test failed: " << ex.what() << "\n";
    rc = 1;
  }

  tamm::finalize(true);
  return rc;
}